daemon/mgr/container.go

```diff
@@ -1175,6 +1179,69 @@ func (mgr *ContainerManager) Restart(ctx context.Context, name string, timeout i
 	return mgr.start(ctx, c, "")
 }
 
+// DisconnectContainerFromNetwork disconnects the given container from
+// given network
+func (mgr *ContainerManager) DisconnectContainerFromNetwork(ctx context.Context, containerName, networkName string, force bool) error {
+	c, err := mgr.container(containerName)
+	if err != nil {
+		// TODO(ziren): if force is true, force delete endpoint
+		return err
+	}
+
+	// Get network
+	network, err := mgr.NetworkMgr.Get(ctx, networkName)
+	if err != nil {
+		return fmt.Errorf("failed to get network %s when disconnecting container %s: %v", networkName, c.Name(), err)
+	}
+
+	// container cannot be disconnected from host network
+	networkMode := c.meta.HostConfig.NetworkMode
+	if IsHost(networkMode) && IsHost(network.Mode) {
+		return fmt.Errorf("container cannot be disconnected from host network or connected to hostnetwork ")
+	}
+
+	networkSettings := c.meta.NetworkSettings
+	if networkSettings == nil {
+		return nil
+	}
+
+	epConfig, ok := networkSettings.Networks[network.Name]
+	if !ok {
+		// container not attached to the given network
+		return fmt.Errorf("failed to disconnect container from network: container %s not attach to %s", c.Name(), networkName)
+	}
+
+	endpoint := mgr.buildContainerEndpoint(c.meta)
+	endpoint.Name = network.Name
+	endpoint.EndpointConfig = epConfig
+	if err := mgr.NetworkMgr.EndpointRemove(ctx, endpoint); err != nil {
+		logrus.Errorf("failed to remove endpoint: %v", err)
+		return err
+	}
+
+	// disconnect an endpoint success, delete endpoint info from container json
+	delete(networkSettings.Networks, network.Name)
+
+	// if container has no network attached any more, set NetworkDisabled to true
+	// so that not setup Network Namespace when restart the container
+	if len(networkSettings.Networks) == 0 {
+		c.meta.Config.NetworkDisabled = true
+	}
+
+	// container meta changed, refresh the cache
+	// remove old container from cache
+	mgr.cache.Remove(c.ID())
+	// add new container to cache
+	mgr.cache.Put(c.ID(), c)
```

Comment:
Removing it and adding it again is our current way to update an instance. 
I think it is kind of redundant. I wish that we could add an Update function in package `satemap`.

Not block this. Just record this.

Reply:
Continue to add a comment to this. I am afraid we do not need to function Update, because we can use `Put` function directly to achieve this. @HusterWan 

Since the code is as following:
```
// Put stores a key-value pair into inner map safely.
func (m *SafeMap) Put(k string, v interface{}) {
	m.Lock()
	defer m.Unlock()

	if m.inner == nil {
		return
	}

	m.inner[k] = v
}
```

So no need to remove, just put is OK.

Reply:
Yes, i will create a new pr to refactor this part of code.